tensorboard/java/org/tensorflow/tensorboard/vulcanize/Vulcanize.java

```diff
@@ -793,22 +802,73 @@ private static void combineScriptElements(Document document) {
         }
         sourceContent = new String(Files.readAllBytes(webfiles.get(webpath)), UTF_8);
       }
-      String hash = BaseEncoding.base64().encode(
-          Hashing.sha256().hashString(sourceContent, UTF_8).asBytes());
+      String hash =
+          BaseEncoding.base64().encode(Hashing.sha256().hashString(sourceContent, UTF_8).asBytes());
       hashes.add(hash);
     }
     return hashes;
   }
 
-  // Writes sha256 of script tags in base64 in the document.
-  private static void writeShasum(Document document, Path output) throws FileNotFoundException, IOException {
-    String hashes = Joiner.on("\n").join(computeScriptShasum(document));
-    Files.write(
-        output,
-        hashes.getBytes(UTF_8),
-        StandardOpenOption.WRITE,
-        StandardOpenOption.CREATE,
-        StandardOpenOption.TRUNCATE_EXISTING);
+  /**
+   * Writes sha256 of script tags in base64 in the document.
+   *
+   * @deprecated Shasum is deprecated in favor of script extraction.
+   */
+  @Deprecated
+  private static String getScriptsShasums(Document document)
+      throws FileNotFoundException, IOException {
+    return Joiner.on("\n").join(computeScriptShasum(document));
+  }
+
+  private static String extractScriptContent(Document document)
+      throws FileNotFoundException, IOException, IllegalArgumentException {
+    Elements scripts = document.getElementsByTag("script");
+    StringBuilder sourcesBuilder = new StringBuilder();
+
+    for (Element script : scripts) {
+      String sourceContent;
+      String src = script.attr("src");
+      if (src.isEmpty()) {
+        sourceContent = script.html();
+      } else {
+        // script element that remains are the ones with src that is absolute or annotated with
+        // `jscomp-ignore`. They must resolve from the root because those srcs are rootified.
+        Webpath webpathSrc = Webpath.get(src);
+        Webpath webpath = Webpath.get("/").resolve(webpathSrc).normalize();
+        if (isAbsolutePath(webpathSrc)) {
+          if (script.hasAttr("defer") || script.hasAttr("async")) {
+            continue;
+          }
+          throw new IllegalArgumentException(
+              "Script refers to a remote resource ("
+                  + webpathSrc
+                  + ") in a blocking way. For"
+                  + " correctness of execution, please make sure it is async-able or defer-able:"
+                  + script.outerHtml());
+        } else if (!webfiles.containsKey(webpath)) {
+          throw new FileNotFoundException(
+              "Expected webfiles for " + webpath + " to exist. Related: " + script.outerHtml());
+        }
+        sourceContent = new String(Files.readAllBytes(webfiles.get(webpath)), UTF_8);
+      }
+
+      sourcesBuilder.append(sourceContent).append("\n");
```

Comment:
Should we include a comment here indicating the provenance of the file,
for debugging purposes? Or are the `~~~WEBPATH~~~` (or whatever)
comments included here already?

Reply:
Thinking out loud here.

This is combining already inlined scripts so they do not have `webpath` or name associated with them. We can put indices but I am not sure how useful that is. Also, while compiling, we've already put bunch of `sourceURL=` for souremap. What kind of delimitation would help us debug issues?

Reply:
Okay, source map URLs should be sufficient. Thanks for considering!